Feature/linking

fbacker opened the pull request:
Having quite big spec files so would be nice to allow collapse/expand.

This PR allows to collapse tags. They are default expanded and if they should be collapsed and search is used they will expand to show all results.

Tried a couple of examples to try it out and seems to work. Couldn't see the same use case for 'example5.html' to collapse the side navigation so didn't build support for it.

Found this talking about the feature https://github.com/mrin9/RapiDoc/issues/127

mrin9 commented:
Thanks for the PR !!!
I haven't tested it but a quick look through makes me think it break markdown / internal links  
for instance this link 
https://mrin9.github.io/RapiDoc/examples/example1.html#post-/user should take me to the desired operation with in the spec

but can you confirm me if it will work the same when tags are collapsed and you change the location hash in the url that points to an operation inside the collapsed tag?

fbacker commented:
by default all tags are expanded. So internal linking when loading a page isn't a problem.

however I've added the possibility (currently needs PR https://github.com/mrin9/RapiDoc/pull/172) to use the tags in schema that can be used for sorting. They can also be used to collapse a tag by default. Then this would't work.

Thinking that is a special case (that I'm currently using) when creator wants to collapse tags at start, then it might as a start be okay that internal linking to that operation doesn't work. Think it could be easy to fix by finding operation on boot and setting the parent tag to exapanded = true. But could also wait until someone needs that functionality.

fbacker commented:
I can make a quick look to auto expand a collapsed on hash change.

fbacker commented:
on load collapsed tags are opened. added example 'collapsable.html'

mrin9 commented:
>by default all tags are expanded. So internal linking when loading a page isn't a problem.

I saw you added `expanded` attribute to `tags` which is good, we want the design to be declarative as much as possible and it is fine to not expose that attribute out and have a default, 
But still it should support both the values `expanded:true` or `expanded:false`.  In this case it wont work if I default all the tags to `expanded:false`. This feature is specially useful for larger specs and most of them would be wanting to show the spec in collapsed form

>however I've added the possibility (currently needs PR #172) to use the tags in schema that can be used for sorting

This functionality already exist https://mrin9.github.io/RapiDoc//examples/arrange-by-tags.html#overview
I haven't looked at your PR but I hope you are not re-doing/changing/breaking any of the existing functionality ?

> I can make a quick look to auto expand a collapsed on hash change.

I just gave you an example of URL change on location bar, but there are other ways people provide navigation functionality in the spec such as
- creating markdown links inside the spec to other paths
- uses `scrollTo(path)` [method] (https://mrin9.github.io/RapiDoc/api.html#methods)

fbacker commented:
> But still it should support both the values `expanded:true` or `expanded:false`.

Yes. If no extra ```tags``` information is given. The value will be ```true```. Just thought everything went with tag default value creation. Fixed now.

> however I've added the possibility (currently needs PR #172) to use the tags in schema that can be used for sorting

Bad description of me. I've used the same way that you created the sorting ``´tags``` in schema. Just added the extra possibility to put expanded here. Se example of collapsable.html

Added fixes and testing for ```markdown link``` and ```scrollTo(path)``` in the example.

mrin9 commented:
I wont be able to test it, unless you resolve the conflicts

fbacker commented:
Quite a busy master branch so hard to keep up-to-date with my PRs.

mrin9 commented:
no issues, take your time ,  we are getting ready for an release so wanted to test if it can be included

fbacker commented:
Gr8. Don’t know if you saw but made push when I wrote that doesn’t have
conflicts.

On Wed, 8 Apr 2020 at 23:54, Mrinmoy <notifications@github.com> wrote:

> no issues, take your time , we are getting ready for an release so wanted
> to test if it can be included
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mrin9/RapiDoc/pull/171#issuecomment-611214423>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCFLGKXEKG4UTBVEORFFSLRLTXBDANCNFSM4L2UFJ7Q>
> .
>
-- 
Mvh,
Fredrick Bäcker

mrin9 commented:
Thanks it was a long standing and popular requested feature !!!

reyolivo commented:
Hello, thanks for this contribution and solution. I do see this doesn't work for the read/focused mode. Is there any way to do this in those render modes?
